pkg/cmd/podidentity/detect.go

```diff
@@ -0,0 +1,410 @@
+					if ownerReference.Controller != nil && *ownerReference.Controller {
+						ownerReferences[ownerReference] = azureIdentity.Spec.ClientID
+						ownerFound = true
+						break
+					}
+				}
+			}
+			// this is a standalone pod, so add it to the results
+			if !ownerFound {
+				p := pods[i]
+				results[&p] = azureIdentity.Spec.ClientID
+			}
+		}
+	}
+
+	for ownerReference, clientID := range ownerReferences {
+		owner, err := dc.getOwner(ownerReference)
+		if err != nil {
+			return err
+		}
+		results[owner] = clientID
+	}
+
+	// results contains all the resources that we need to generate a config file.
+	// for each entry in the results map, we will generate a service account yaml file
+	// and a resource file
+	for o, clientID := range results {
+		localObject := k8s.NewLocalObject(o)
+
+		sa, err := dc.createServiceAccountFile(localObject.GetServiceAccountName(), localObject.GetName(), clientID)
+		if err != nil {
+			return err
+		}
+		if err = dc.createResourceFile(localObject, sa); err != nil {
+			return err
+		}
+		log.Debugf("generated config for %s/%s, clientID: %s", strings.ToLower(localObject.GetObjectKind().GroupVersionKind().Kind), localObject.GetName(), clientID)
+	}
+
+	if len(results) == 0 {
+		log.Debugf("no aad-pod-identity configuration found in namespace: %s", dc.namespace)
+		return nil
+	}
+
+	log.Infof("generated resource and service account files in output directory: %s\n%s", dc.outputDir, nextStepsLogMessage)
+	return nil
+}
+
+// createServiceAccountFile will create a service account yaml file
+// 1. If the resource is using default service account, then a new service account yaml is generated
+//    with the resource name as service account name
+// 2. If the resource is already using a non-default service account, then we modify that service account
+//    to generate the desired yaml file
+// The service account yaml will contain the workload identity use label ("azure.workload.identity/use: true")
+// and the client-id annotation ("azure.workload.identity/client-id: <client-id from AzureIdentity>")
+func (dc *detectCmd) createServiceAccountFile(name, ownerName, clientID string) (*corev1.ServiceAccount, error) {
+	sa := &corev1.ServiceAccount{}
+	var err error
+	if name == "" || name == "default" {
+		log.Debugf("%s is using default service account, generating a new service account", ownerName)
+		// generate a new service account yaml file with owner name as service account name
+		sa.SetName(ownerName)
+		sa.SetNamespace(dc.namespace)
+	} else {
+		// get service account referenced by the owner
+		if sa, err = kuberneteshelper.GetServiceAccount(context.TODO(), dc.kubeClient, dc.namespace, name); err != nil {
+			return nil, err
+		}
+	}
+
+	saLabels := make(map[string]string)
+	if sa.GetLabels() != nil {
+		saLabels = sa.GetLabels()
+	}
+	saLabels[webhook.UseWorkloadIdentityLabel] = "true"
+	sa.SetLabels(saLabels)
+
+	// set the annotations for the service account
+	saAnnotations := make(map[string]string)
```

Comment:
Should we also account `azure.workload.identity/service-account-token-expiration` and `azure.workload.identity/tenant-id` in case the user wants to customize them within the same namespace? I am not sure if we can customize `azure.workload.identity/skip-containers` since workloads might not have the same set of skip containers.

Reply:
Great suggestion! Let me add flags for expiration and tenant-id.

I agree we can't provide a good experience for customizing skip containers because that's different across the workloads in the namespace. I was worried about customizing tenant-id too but there is no [cross-tenant option](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/managed-identities-faq#can-i-use-a-managed-identity-to-access-a-resource-in-a-different-directorytenant), so all the identities will be in the same namespace and most likely be in the same tenant as cluster.

Reply:
Done!